AndroidSDKFcm/src/main/java/com/leanplum/LeanplumPushFirebaseMessagingService.java

```diff
@@ -86,4 +99,63 @@ private Bundle getBundle(Map<String, String> messageMap) {
     }
     return bundle;
   }
+
+  private Boolean isDuplicateNotification(Map<String, String> messageMap) {
+    Boolean isDuplicate = false;
+    String messageId = "";
+    if (messageMap.containsKey("_lpn")) {
+      messageId = messageMap.get("_lpn");
+    }
+    if (messageMap.containsKey("_lpm")) {
+      messageId = messageMap.get("_lpm");
+    }
+
+    Map<String, Object> handledNotifications = retrieveHandledNotifications();
+    handledNotifications = purgeExpiredHandledNotifications(handledNotifications);
+    if (handledNotifications.containsKey(messageId)) {
+      isDuplicate = true;
+    } else {
+      handledNotifications.put(messageId, new Date().getTime());
+    }
+    persistHandledNotifications(handledNotifications);
+    return isDuplicate;
+  }
+
+  private Map<String, Object> retrieveHandledNotifications() {
```

Comment:
The plan as per the doc was using VarCache, because it handles encryption and decryption code. Also, does have the same point of responsibility of storing anything caching.